lib/src/executable/watch.dart

```diff
@@ -176,29 +194,38 @@ class _Watcher {
 
   /// Recompiles [nodes] and everything that transitively imports them, if
   /// necessary.
-  Future _recompileDownstream(Iterable<StylesheetNode> nodes) async {
+  ///
+  /// Returns whether all recompilations succeeded.
+  Future<bool> _recompileDownstream(Iterable<StylesheetNode> nodes) async {
     var seen = new Set<StylesheetNode>();
     var toRecompile = new Queue.of(nodes);
 
+    var allSucceeded = true;
     while (!toRecompile.isEmpty) {
       var node = toRecompile.removeFirst();
       if (!seen.add(node)) continue;
 
-      await _compileIfEntrypoint(node.canonicalUrl);
+      var success = await _compileIfEntrypoint(node.canonicalUrl);
+      allSucceeded = allSucceeded && success;
+      if (!success && _options.stopOnError) return false;
+
       toRecompile.addAll(node.downstream);
     }
+    return allSucceeded;
   }
 
   /// Compiles the stylesheet at [url] to CSS if it's an entrypoint that's being
   /// watched.
-  Future _compileIfEntrypoint(Uri url) async {
-    if (url.scheme != 'file') return;
+  ///
+  /// Returns `false` if compilation failed, `true` otherwise.
+  Future<bool> _compileIfEntrypoint(Uri url) async {
+    if (url.scheme != 'file') return true;
 
     var source = p.fromUri(url);
     var destination = _destinationFor(source);
-    if (destination == null) return;
+    if (destination == null) return true;
 
-    await compile(source, destination);
+    return await compile(source, destination);
```

Comment:
`return compile(source, destination);`?  (slightly faster/less chars)

(same below)

Reply:
I actually strongly prefer `return await`, because it makes the function much more resilient against refactors. If you just write `return`, it's easy to miss that the function you're returning is asynchronous when you come back later, which can lead to dangling futures.

Reply:
can you explain more what you mean by this? what kind of refactor are you envisioning?

Reply:
Suppose I wanted to add error handling, for example. If I had just written `return compile(...)`, it would be easy to accidentally write:

```dart
try {
  return compile(source, destination);
} catch (error) {
  // ...
}
```

But this code is buggy; it won't actually catch exceptions thrown asynchronously by `compile()`. `return await` avoids that issue.

Reply:
interesting. Seems suboptimal to add `await` everywhere because we might hypothetically add a try+catch to handle async errors around a return. But, up to you.

Reply:
I generally try to write code in a way that requires future edits to re-load as little knowledge as possible into their heads about what all the code is doing. I think that's worth a few extra tokens. As for runtime performance, I don't think this makes a measurable difference, and if it ever became a serious problem it would be easy to optimize away at the compiler level.